src/main/scala/uk/gov/ons/registers/methods/PayeCalculator.scala

```diff
@@ -0,0 +1,152 @@
+package uk.gov.ons.registers.methods
+
+import org.apache.spark.sql.functions.{explode_outer, sum}
+import org.apache.spark.sql.{DataFrame, SparkSession}
+
+
+trait PayeCalculator {
+
+  val jobs = "paye_jobs"
+  val employees = "paye_empees"
+
+  def calculatePAYE(BIDF: DataFrame, payeDF: DataFrame)(implicit activeSession: SparkSession): DataFrame = {
+    val calculatedPayeEmployeesDF = getGroupedByPayeEmployees(BIDF, payeDF)
+    val calculatedPayeJobsDF = getGroupedByPayeJobs(BIDF, payeDF, "dec_jobs")
+    val calculatedPayeDF = calculatedPayeEmployeesDF.join(calculatedPayeJobsDF, "ern")
+    //calculatedPayeDF.show
+
+    calculatedPayeDF
+  }
+
+  def getGroupedByPayeEmployees(BIDF: DataFrame, payeDF: DataFrame, luTableName: String = "LEGAL_UNITS", payeDataTableName: String = "PAYE_DATA")(implicit spark: SparkSession): DataFrame ={
+    val flatUnitDf = BIDF.withColumn("payeref", explode_outer(BIDF.apply("PayeRefs")))
```

Comment:
If we were to change the column name of a key variable such as "payeref" - there are multiple places in this code that will need to be tracked down and changed.

This is why we normally define constants for key column names, so that if the database representation changes, the column name only needs to be changed at **one** place in the code.
We don't need constants for all the columns, but key columns that are used for joins (and are therefore referenced in multiple places) should be defined that way.